exampleCustomAudioDevice/src/main/java/com/twilio/examplecustomaudiodevice/CustomAudioDeviceActivity.kt

```diff
@@ -0,0 +1,428 @@
+            /*
+             * OPTION 1 - Generate an access token from the getting started portal
+             * https://www.twilio.com/console/video/dev-tools/testing-tools and add
+             * the variable TWILIO_ACCESS_TOKEN setting it equal to the access token
+             * string in your local.properties file.
+             */
+            accessToken = TWILIO_ACCESS_TOKEN
+        } else {
+            /*
+             * OPTION 2 - Retrieve an access token from your own web app.
+             * Add the variable ACCESS_TOKEN_SERVER assigning it to the url of your
+             * token server and the variable USE_TOKEN_SERVER=true to your
+             * local.properties file.
+             */
+            retrieveAccessTokenfromServer()
+        }
+    }
+
+    private fun connectToRoom(roomName: String) {
+        configureAudio(true)
+        val connectOptionsBuilder = ConnectOptions.Builder(accessToken!!)
+                .roomName(roomName)
+
+        /*
+         * Add local audio track to connect options to share with participants.
+         */if (localAudioTrack != null) {
+            connectOptionsBuilder
+                    .audioTracks(listOf(localAudioTrack))
+        }
+        room = Video.connect(this, connectOptionsBuilder.build(), roomListener())
+        setDisconnectAction()
+    }
+
+    private fun retrieveAccessTokenfromServer() {
+        Ion.with(this)
+                .load(String.format("%s?identity=%s", ACCESS_TOKEN_SERVER,
+                        UUID.randomUUID().toString()))
+                .asString()
+                .setCallback { e: Exception?, token: String? ->
+                    if (e == null) {
+                        accessToken = token
+                    } else {
+                        Toast.makeText(this@CustomAudioDeviceActivity,
+                                R.string.error_retrieving_access_token, Toast.LENGTH_LONG)
+                                .show()
+                    }
+                }
+    }
+
+    private fun checkPermissionForMicrophone(): Boolean {
+        val resultMic = ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)
+        return resultMic == PackageManager.PERMISSION_GRANTED
+    }
+
+    private fun requestPermissionForMicrophone() {
+        if (ActivityCompat.shouldShowRequestPermissionRationale(this,
+                        Manifest.permission.RECORD_AUDIO)) {
+            Toast.makeText(this,
+                    R.string.permissions_needed,
+                    Toast.LENGTH_LONG).show()
+        } else {
+            ActivityCompat.requestPermissions(
+                    this, arrayOf(Manifest.permission.RECORD_AUDIO),
+                    MIC_PERMISSION_REQUEST_CODE)
+        }
+    }
+
+    private fun createAudioTrack() {
+        // Share your microphone
+        localAudioTrack = LocalAudioTrack.create(this, true, LOCAL_AUDIO_TRACK_NAME)
+    }
+
+    private fun configureAudio(enable: Boolean) {
```

Comment:
Remove usage of `!!`. This can usually be replaced with an `?` operator and the use of `let` or `apply`